stevenelleman/sg
# API Service

## Terms

## What 
Create API service.

## How 
Request: 
`API Router` -request-> `Handler` -req-> `Controller` -req-> `DBI` -req-> `Postgres DB`  
Response: 
... the other way ... 

## Why 
Gotta interact with the database. 

## Tasks 

## Notes to Reviewers 
There are lots of improvements to be made (testing, per-request db connections, local table db loading, authorization context, linter, etc) but this _does_ work. 

## How to Test
To test this you'll need to setup postgres. Make a new db called `stevenelleman`. After that these should be the steps to test the service: 
- `brew install go`
- (in `sg` directory) `git fetch origin`
- `git checkout se/api`
- `cd services/public-api`
- `make run`  will run the service + db 

Manually add the silo table to psql:
- `psql stevenelleman`
- `CREATE TABLE silos (id varchar(255), state varchar(255));`
- `\q` -> exit psql 

Play with silo rows in the DB: 
- Make new silo: `curl -X POST localhost:8000/v1/silos/${random id}` -> create silo with the `random id`
- Get silo: `curl localhost:8000/v1/silos/${random id}`
- List silos: `curl localhost:8000/v1/silos`
- Delete silo: `curl -X DELETE localhost:8000/v1/silos/${random id}`

## Meta
